flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -795,20 +806,241 @@ private void restoreInstance(
+								} else {
+									// Since the iterator will visit the record according to the sorted order,
+									// we can just break here.
+									break;
+								}
+
+								iterator.next();
+							}
+						} // releases native iterator resources
+					}
+				} finally {
+					FileSystem restoreFileSystem = temporaryRestoreInstancePath.getFileSystem();
+					if (restoreFileSystem.exists(temporaryRestoreInstancePath)) {
+						restoreFileSystem.delete(temporaryRestoreInstancePath, true);
+					}
+				}
+			}
+		}
+
+		private class RestoredDBInstance implements AutoCloseable {
+
+			@Nonnull
+			private final RocksDB db;
+
+			@Nonnull
+			private final ColumnFamilyHandle defaultColumnFamilyHandle;
+
+			@Nonnull
+			private final List<ColumnFamilyHandle> columnFamilyHandles;
+
+			@Nonnull
+			private final List<ColumnFamilyDescriptor> columnFamilyDescriptors;
+
+			@Nonnull
+			private final List<RegisteredKeyedBackendStateMetaInfo.Snapshot<?, ?>> stateMetaInfoSnapshots;
+
+			public RestoredDBInstance(@Nonnull RocksDB db,
+								@Nonnull List<ColumnFamilyHandle> columnFamilyHandles,
+								@Nonnull List<ColumnFamilyDescriptor> columnFamilyDescriptors,
+								@Nonnull List<RegisteredKeyedBackendStateMetaInfo.Snapshot<?, ?>> stateMetaInfoSnapshots) {
+				this.db = db;
+				this.columnFamilyHandles = columnFamilyHandles;
+				this.defaultColumnFamilyHandle = this.columnFamilyHandles.remove(0);
+				this.columnFamilyDescriptors = columnFamilyDescriptors;
+				this.stateMetaInfoSnapshots = stateMetaInfoSnapshots;
+			}
+
+			@Override
+			public void close() throws Exception {
+
+				IOUtils.closeQuietly(defaultColumnFamilyHandle);
+
+				for (ColumnFamilyHandle columnFamilyHandle : columnFamilyHandles) {
+					IOUtils.closeQuietly(columnFamilyHandle);
+				}
+
+				IOUtils.closeQuietly(db);
+			}
+		}
+
+		private RestoredDBInstance restoreDBFromStateHandle(
+			IncrementalKeyedStateHandle restoreStateHandle,
+			Path temporaryRestoreInstancePath,
+			KeyGroupRange targetKeyGroupRange,
+			int keyGroupPrefixBytes,
+			boolean needClip) throws Exception {
+
+			transferAllStateDataToDirectory(restoreStateHandle, temporaryRestoreInstancePath);
+
 			// read meta data
 			List<RegisteredKeyedBackendStateMetaInfo.Snapshot<?, ?>> stateMetaInfoSnapshots =
-				readMetaData(localKeyedStateHandle.getMetaDataState());
+				readMetaData(restoreStateHandle.getMetaStateHandle());
 
 			List<ColumnFamilyDescriptor> columnFamilyDescriptors =
 				createAndRegisterColumnFamilyDescriptors(stateMetaInfoSnapshots);
 
-			restoreLocalStateIntoFullInstance(
-				localKeyedStateHandle,
+			List<ColumnFamilyHandle> columnFamilyHandles =
+				new ArrayList<>(stateMetaInfoSnapshots.size() + 1);
+
+			RocksDB restoreDb = stateBackend.openDB(
+				temporaryRestoreInstancePath.getPath(),
 				columnFamilyDescriptors,
-				stateMetaInfoSnapshots);
+				columnFamilyHandles);
+
+			if (needClip) {
```

Comment:
Instead of using this clipping as a flag, would it not be better to just have a method that clips a `RestoredDBInstance` which is simply called after `restoreDBFromStateHandle` in the on case that needs it? This would allow to not mix up those two tings in one method.

Reply:
In that case, maybe `clip` can also become a method of the `RestoredDBInstance`?

Reply:
Yes, you are right, will follow your suggestion.